mlflow/evaluation/utils.py

```diff
@@ -0,0 +1,188 @@
+
+
+def evaluations_to_dataframes(
+    evaluations: List[EvaluationEntity],
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """
+    Converts a list of Evaluation entities to four separate DataFrames: one for main evaluation
+    data (excluding assessments and metrics), one for metrics, one for assessments, and one for
+    tags.
+
+    Args:
+        evaluations (List[Evaluation]): List of Evaluation entities.
+
+    Returns:
+        Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]: A tuple of four DataFrames
+        containing evaluation data, metrics data, assessments data, and tags data.
+    """
+    evaluations_data = []
+    metrics_data = []
+    assessments_data = []
+    tags_data = []
+
+    for evaluation in evaluations:
+        eval_dict = evaluation.to_dictionary()
+
+        # Extract assessment and metrics
+        assessments_list = eval_dict.pop("assessments", [])
+        metrics_list = eval_dict.pop("metrics", [])
+        tags_list = eval_dict.pop("tags", [])
+
+        evaluations_data.append(eval_dict)
+
+        for metric_dict in metrics_list:
+            metric_dict["evaluation_id"] = eval_dict["evaluation_id"]
+            # Remove 'step' key if it exists, since it is not valid for evaluation metrics
+            metric_dict.pop("step", None)
+            metrics_data.append(metric_dict)
+
+        for assess_dict in assessments_list:
+            assess_dict["evaluation_id"] = eval_dict["evaluation_id"]
+            assessments_data.append(assess_dict)
+
+        for tag_dict in tags_list:
+            tag_dict["evaluation_id"] = eval_dict["evaluation_id"]
+            tags_data.append(tag_dict)
+
+    evaluations_df = (
+        _apply_schema_to_dataframe(
+            pd.DataFrame(evaluations_data), _get_evaluations_dataframe_schema()
+        )
+        if evaluations_data
+        else _get_empty_evaluations_dataframe()
+    )
+    metrics_df = (
+        _apply_schema_to_dataframe(pd.DataFrame(metrics_data), _get_metrics_dataframe_schema())
+        if metrics_data
+        else _get_empty_metrics_dataframe()
+    )
+    assessments_df = (
+        _apply_schema_to_dataframe(
+            pd.DataFrame(assessments_data), _get_assessments_dataframe_schema()
+        )
+        if assessments_data
+        else _get_empty_assessments_dataframe()
+    )
+    tags_df = (
+        _apply_schema_to_dataframe(pd.DataFrame(tags_data), _get_tags_dataframe_schema())
+        if tags_data
+        else _get_empty_tags_dataframe()
+    )
+
+    return evaluations_df, metrics_df, assessments_df, tags_df
+
+
+def _get_evaluations_dataframe_schema() -> Dict[str, str]:
+    """
+    Returns the pandas schema for the evaluation DataFrame.
+    """
+    return {
+        "evaluation_id": "string",
+        "run_id": "string",
+        "inputs_id": "string",
+        "inputs": "object",
+        "outputs": "object",
+        "request_id": "object",
```

Comment:
Optional fields are represented as "object" type, even though `request_id` is technically Optional[str].

The reason for using "object" instead of "string" is that pandas does not respect the "string" type when reading a pandas dataframe from JSON if all of the values are `null`. I've seen cases where pandas implicitly casts the column to a numerical value type, etc.

Empirically, `object` seems to work properly.

Reply:
I think that's the only way to make Pandas treat an entry "as-is" +1